language-support/ts/daml-react/hooks.test.ts

```diff
@@ -200,3 +232,138 @@ describe('useFetchByKey', () => {
     expect(result.current.queryResult).toEqual({contract, loading: false});
   });
 });
+
+describe('useStreamQuery', () => {
+  test('empty stream', () => {
+    // setup
+    const query = 'foo-query';
+    const [stream, emitter] = mockStream();
+    mockStreamQuery.mockReturnValueOnce(stream);
+    const hookResult = renderDamlHook(() => useStreamQuery(Foo, () => ({query}), [query]));
+    expect(mockStreamQuery).toHaveBeenCalledTimes(1);
+    expect(mockStreamQuery).toHaveBeenLastCalledWith(Foo, {query: query});
+    expect(hookResult.result.current.contracts).toEqual([]);
+    expect(hookResult.result.current.loading).toBe(false);
+
+    // no events have been emitted.
+    expect(hookResult.result.current.contracts).toEqual([]);
+    expect(hookResult.result.current.loading).toBe(false);
+
+    // empty events
+    act(() =>
+      void(emitter.emit('change', []))
+    );
+    expect(hookResult.result.current.contracts).toEqual([]);
+    expect(hookResult.result.current.loading).toBe(false);
+  });
+
+  test('new events', () => {
+    // setup
+    const query = 'foo-query';
+    const [stream, emitter] = mockStream();
+    mockStreamQuery.mockReturnValueOnce(stream);
+    const hookResult = renderDamlHook(() => useStreamQuery(Foo, () => ({query}), [query]));
+    expect(mockStreamQuery).toHaveBeenCalledTimes(1);
+    expect(mockStreamQuery).toHaveBeenLastCalledWith(Foo, {query: query});
+    expect(hookResult.result.current.contracts).toEqual([]);
+    expect(hookResult.result.current.loading).toBe(false);
+
+    // one new event
+    act(() => void(emitter.emit('change', ['event1'])));
+    expect(hookResult.result.current.contracts).toEqual(['event1']);
+    expect(hookResult.result.current.loading).toBe(false);
+
+    // two new events replacing old one.
+    act(() => void(emitter.emit('change', ['event2', 'event3'])));
+    expect(hookResult.result.current.contracts).toEqual(['event2', 'event3']);
+    expect(hookResult.result.current.loading).toBe(false);
+  });
+
+  test('changed query triggers reload', () => {
+    // setup
+    const query1 = 'foo-query';
+    const query2 = 'bar-query';
+    const [stream, emitter] = mockStream();
+    mockStreamQuery.mockReturnValueOnce(stream);
+    const {result} = renderDamlHook(() => {
+      const [query, setQuery] = useState(query1);
+      const queryResult = useStreamQuery(Foo, () => ({query}), [query]);
+      return {queryResult, query, setQuery};
+    })
+    expect(mockStreamQuery).toHaveBeenCalledTimes(1);
+    expect(mockStreamQuery).toHaveBeenLastCalledWith(Foo, {query: query1});
+    expect(result.current.queryResult.contracts).toEqual([]);
+    expect(result.current.queryResult.loading).toBe(false);
+
+    // new events
+    act(() => void(emitter.emit('change', ['foo'])));
+    expect(result.current.queryResult.contracts).toEqual(['foo']);
```

Comment:
Can we please always check for the `loading` indicator as well. Currently, this indicator is not super useful but this will change. And having a good test infrastructure in place for that would be helpful.

Reply:
changed.